\name{bismark2segment}
\alias{bismark2segment}
%- Also NEED an '\alias' for EACH other topic documented here.
\title{
%%  ~~function to do ... ~~
Process bismark report file to 4-CpG segments information for pCSM loci identification
}
\description{
%%  ~~ A concise (1-5 lines) description of what the function does. ~~
This function is used for processing data into the format that could be recognised by beta mixture model or nonparametric Bayesian clustering algorithm to identify pCSM loci.
}
\usage{
bismark2segment(files,file_type="regular",CpG_file,tmp_folder="temp",split_by_chrom=FALSE)
}
%- maybe also 'usage' for other objects documented here.
\arguments{
  \item{files}{File or files with CpG methylation information in each sequenced read generated by bismark_extractor in ".gz" compressed format. Note that only one filename with full path is needed for regular methylation dataset and a vector containning the filename with full path of each single-cell is needed for processing single-cell datasets.
%%     ~~Describe \code{x} here~~
}
  \item{file_type}{Type of input dataset with "regular" represents the regular methylation data and "single-cell" represents single-cell mathylation data.
%%     ~~Describe \code{x} here~~
}
  \item{CpG_file}{File includes the coordinate of all the CpG loci in forward strand of the target genome, with two columns separated by tab, the chromsome id in the first column and location in the second column
%%     ~~Describe \code{x} here~~
}
\item{split_by_chrom}{Logical; Used for single-cell datasets when the number of cells is huge. Note that by setting split_by_chrom=TRUE, a list will be returned with each elements represents the input of beta mixture model for one chromsome
%%     ~~Describe \code{x} here~~
}

}
\details{
%%  ~~ If necessary, more details than the description above ~~
}
\value{
%%  ~Describe the value returned
%%  If it is a LIST, use
%%  \item{comp1 }{Description of 'comp1'}
%%  \item{comp2 }{Description of 'comp2'}
%% ...
\item{segment}{A matrix or a list containning the 4CpG segments infromation for pCSM loci identification. }
}
\references{
%% ~put references to the literature/web site here ~
}
\author{
%%  ~~who you are~~
}
\note{
%%  ~~further notes~~
loading and processing the CpG index may need several minutes
}

%% ~Make other sections like Warning with \section{Warning }{....} ~

\seealso{
%% ~~objects to See Also as \code{\link{help}}, ~~~
}
\examples{
library('csmFinder')
#For bulk methylome, the 4-CpG segments could be extracted as follow:

#get the demo datasets
bismark_result=paste(system.file(package="csmFinder"),"extdata/bulk_CpG_extract_file/demo.dataset.gz",sep='/')
CpG_ref=paste(system.file(package="csmFinder"),"extdata/CpG_plus.reference",sep='/')

#generate the 4-CpG segment
segment <- bismark2segment(files=bismark_result,CpG_file=CpG_ref)

#see what the segment looks like
segment[2:5,] 
 


#############################
#For single-cell methylomes, file_type="single-cell" argument is needed, and the 4-CpG segments could be extracted as follow:

#get the demo datasets
scDataDir <- paste(system.file(package="csmFinder"),"extdata/single_cell_CpG_extract_file",sep='/')
file_list <- paste(scDataDir,list.files(scDataDir),sep='/')

#generate the 4-CpG segment
scSegment <- bismark2segment(files=file_list,file_type="single-cell",CpG_file=CpG_ref)

}
% Add one or more standard keywords, see file 'KEYWORDS' in the
% R documentation directory.
\keyword{ ~kwd1 }% use one of  RShowDoc("KEYWORDS")
\keyword{ ~kwd2 }% __ONLY ONE__ keyword per line
