\name{csmFinder}
\alias{csmFinder}
%- Also NEED an '\alias' for EACH other topic documented here.
\title{
%%  ~~function to do ... ~~
Find cell-subset specific methylation loci in genome.
}
\description{
%%  ~~ A concise (1-5 lines) description of what the function does. ~~
Find cell-subset methylation segments from methylomes generated by single cells or bulk tissue. For single-cell methylomes, a beta mixture model is used for divide the cells into two subsets with hypo and hyper-methylation states in candidate CSM regions. For regular datasets generated from bulk tissue or sorted cell populations, a Nonparametric Bayesian clustering is used to group the sequencing reads into hypo and hyper-methylated subsets, both of the two algorithms identify the 4-CpG segments with biplolar methylation patterns across two subsets as the pCSM loci.
}
\usage{
csmFinder(candidate,data_type='regular',depth=10,
          distance=0.3,pval=0.05,thread=1)
}
%- maybe also 'usage' for other objects documented here.
\arguments{
  \item{candidate}{ the candidate segments used for CSM identification
%%     ~~Describe \code{x} here~~
}
  \item{data_type}{ "regular" and "single-cell" represents regular datasets and single-cell datasets, respectively
%%     ~~Describe \code{x} here~~
}
  \item{depth}{ numeric threshold represents the least number of reads (for regular datasets) or cells (for single-cell datasets) covered the candidate segments
%%     ~~Describe \code{x} here~~
}
  \item{distance}{ methylation difference between hypo and hyper-methylated cells subsets or reads
%%     ~~Describe \code{x} here~~
}
  \item{pval}{ significance of the differnece between hypo and hyper-methylated cells subsets or reads
%%     ~~Describe \code{x} here~~
}
  \item{thread}{
%%     ~~Describe \code{x} here~~
number of threads used to identify candidate pCSM segment
}

}
\details{
%%  ~~ If necessary, more details than the description above ~~
}
\value{
%%  ~Describe the value returned
%%  If it is a LIST, use
%%  \item{comp1 }{Description of 'comp1'}
%%  \item{comp2 }{Description of 'comp2'}
%% ...
For single-cell datasets, the output is in the same format with the output of beta mixture model (https://github.com/Evan-Evans/Beta-Mixture-Model). For regular methylomes, the output is a matrix contains the methylation difference between hypo and hyper-methylated reads, and its significance.
}
\references{
%% ~put references to the literature/web site here ~
Wu, X., et al., 2015, Nonparametric Bayesian clustering to detect bipolar methylated genomic loci, BMC Bioinformatics, 16.

Luo, Y., et al., 2018, Integrative single-cell omics analyses reveal epigenetic heterogeneity in mouse embryonic stem cells, PLoS computational biology, 14, e1006034

}
\author{
%%  ~~who you are~~
}
\note{
%%  ~~further notes~~
}

%% ~Make other sections like Warning with \section{Warning }{....} ~

\seealso{
%% ~~objects to See Also as \code{\link{help}}, ~~~
}
\examples{
###need object for 'candidate' from former steps.  
#for bulk methylome
#pcsm_segment <- csmFinder(candidate,data_type='regular',thread=1)

#for single-cell methylome
#scPcsm_segment <- csmFinder(scCandidate,data_type='single-cell',thread=1)
}

